### PR TITLE
Created search by Genre functionality for Venue page.

### PR DESCRIPTION
**Pull request by lxdnz254**
Added tests for search function.
Moved gulp packages to devDependencies

### CODE REVIEW

**Comment by ttocskcaj** [comment]
Looks good. Only thing is perhaps the api url for searching should be: 
`/api/search/{venue|genre|artist|user}/{searchQuery}`

**Comment by lxdnz254** [comment]
We'll work on that. It was more about me working through the rerouting and
adding a new function, but I see what your asking

**Comment by lxdnz254** [comment]
Thinking a bit more about the api url. It will need to be extended to
/api/search/(venue | band)/(search_attribute e.g. genre)/{searchQuery}
I'm quite happy about implementing our search engine that way. How does it
sit with you?

On 31/07/2017 3:27 PM, "Jack Scott" <notifications@github.com> wrote:

Looks good. Only thing is perhaps the api url for searching should be:
/api/search/{venue|genre|artist|user}/{searchQuery}

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/OpenPolytechnicBITProjectGroup/Musician-Venue_App/pull/26#issuecomment-318958654>,
or mute the thread

**Comment by ttocskcaj** [comment]
When you say 
`GET /api/search/(venue | band)/(search_attribute e.g. genre)/{searchQuery}`
Would that be to search say for a band but filtered by genre? EG:
`GET /api/search/artist/metal/met`
would return Metallica (metal) but not Metronomy (electronic)

I think the correct way to do it would probably be query first then filters (since they should be optional)
`GET /api/search/{artist|venue}/{searchQuery}/{queryFilters}`
`GET /api/search/artist/met?genre=metal&location=Auckland`  (all metal artists in auckland containing "met")
`GET /api/search/artist/met`  (all artists containing "met")
`GET /api/search/venue/?minCapacity=500&location=Christchurch`  (All venues in chch with a capacity => 500. Note the empty search)

Spotify's web API is a good reference: https://developer.spotify.com/web-api/search-item/

Also, just to note something I noticed. We should stick to the word "artist" rather than "band". It's more inclusive as solo artists aren't really a band.

**Comment by lxdnz254** [comment]
I agree with "artist" I'll try change my stream of consciousness towards
that. 😊.
With searching there will be different search criteria.. how do you know
that GET /api/search/artist/metal/met is searching for metal genre? What if
the user wanted to search by cost? eg. GET /api/search/venue/cost>650

I don't think our users will be specifically searching for a name ,so "GET
/api/search/artist/met?genre=metal&location=Auckland (all metal artists in
auckland containing "met")", would be an irrelevant search.
The way I see our app is that a user will log on and based on their
preferences, we will *reccommend *the best options for them.

**Comment by ttocskcaj** [comment]
That's why I suggested 
`GET /api/search/artist/met?genre=metal&location=Auckland`
Perhaps better would be to have the search query as part of the query string as well. Basically like that spotify link I posted.

Get all metal artists in Auckland
`GET /api/search/artist?genre=metal&location=Auckland`

Get all indie artists containing "mySearch"
`GET /api/search/artist?query=mySearch`&genre=Indie

Get any artists containing "mySearch" but limited to Dunedin
`GET /api/search/artist?query=mySearch`&location=Dunedin

One of the important things with designing APIs is that urls shouldn't be changed as it may break any apps using them, hence why it's a good idea to decide on how each URL should work from the start. 
I think it's important to have the ability to add searches for artists by name even if it's not used yet, it will likely be used in future releases.  Using the above, we can just ignore the query parameter until it is needed and only process the location and genre parameters. If a time comes when we need to search for artists by name, adding processing for the query parameter won't break parts of the app that already use the genre or location parameters.

Some examples where searching by name might be useful:

Bob already knows the artists "The Exponents" but want's to find similar artists and in turn, find out what gigs they have coming up. They would search for "The Exponents", and select the artist's profile page. From there, the app could suggest a selection of similar artists.

Sally is a massive fan of "The Exponents" but lives in a small town where they have never played a gig. He goes to his local pub and asks the manager to try to organise an event. The pub's manager would have to search for the band by name to contact them and request a gig.

**Comment by lxdnz254** [comment]
That makes a bit more sense to me, so long as we stay clear on how we are
passing the search queries and filters from client to server, then writing
the appropriate transaction to the database will be a lot simpler.
Remember that our focus for this iteration is to connect artists and
venues... fans are an after thought

**Comment by ttocskcaj** [comment]
Yeah I get that fans are an afterthought, but if/when we add them, we don't want to have to change how the existing parts work. The way I think about it is that the client makes a request to the API to CRUD a resource. That request may or may not have parameters/queries. 

For example, to create an artist resource, a request is made like so:
`POST /api/artists` 
This request would have parameters with the artist's attributes and also some form of authentication. 

To get a specific artist resource (ID of 6):
`GET /api/artists/6` 
There are no parameters needed as the ID represents the resource

To delete a specific artist resource (ID of 6):
`DELETE /api/artists/6' 
As above, no parameters are needed to identify the artist to delete, but there should be some for authentication

Searches should probably work similarly. The search results can be viewed as a resource (just like a list of all artists would be a resource). The only difference is that to get this resource, some parameters may or may not be needed.
`GET /api/search`
This needs parameters such as what to search for and how to narrow it down. Since it's a GET request, the only way to send those parameters is in the URL query string like so:
`GET /api/search?subject=venue&minCapacity=600`

Long story short, the URL should be pointing to the resource required. The query string should specify how to change/manipulate that resource. 

One of the benefits is that it means we can add new filters to the search system without affecting how existing ones work. For example, it's easy to add "numberOfFireExits=9" to the query string that the client requests and then check for it server side, only processing it if it's there.

**Comment by ttocskcaj** [comment]
How easy is it to create a base query for the database and then modify it, based on what search parameters are present?

**Comment by lxdnz254** [comment]
But the api still needs to call a function to the query the database and
return the correct objects to the client.
I'll have a quick play with refactoring the code to be more inline with
this discussion. I think I have a solution mapped out.
Don't want to spend to long though. I have a massive rant coming up on
Facebook about BIT 693 TM1 Q4, that is some NOT "real world entities" going
on and I'm finding it hard to not want to throttle the tutor!

**Comment by lxdnz254** [comment]
That's what I'm trying to map...

**Comment by ttocskcaj** [comment]
For this example:
`GET /api/search?type=venue&genres=[metal,rock]&location=Wellington`
What do you think is more efficient (both in terms of computing and programming time).
Selecting every venue from the database then filtering through those results serverside using node.js

OR

Creating a query that will fetch only what we need from the database. We may have to modify the query depending on what parameters the user is filtering by, possibly with lots of possible combinations of filters

The other possibility is what does neo4j do if part of it's query is blank? For example, the code you've got for searching is:
`MATCH (v:Venue)-[:LIKES_GENRES]->(g:Genre) WHERE (g.name) = {genre} RETURN v AS venue'`
if an empty string for genre is passed into that, would it fail, or return all venues?

**Comment by lxdnz254** [comment]
That would return an empty result. It wouldn't fail as it is a valid
request, but as there's no {genre} nothing is returned.
Basically we would validate client side to make sure valid searches are
sent.

On 31/07/2017 7:53 PM, "Jack Scott" <notifications@github.com> wrote:

For this example:
GET /api/search?type=venue&genres=[metal,rock]&location=Wellington
What do you think is more efficient (both in terms of computing and
programming time).
Selecting every venue from the database then filtering through those
results serverside using node.js

OR

Creating a query that will fetch only what we need from the database. We
may have to modify the query depending on what parameters the user is
filtering by, possibly with lots of possible combinations of filters

The other possibility is what does neo4j do if part of it's query is blank?
For example, the code you've got for searching is:
MATCH (v:Venue)-[:LIKES_GENRES]->(g:Genre) \ WHERE (g.name) = {genre} \
RETURN v AS venue', { genre: genre }
if an empty string for genre is passed into that, would it fail, or return
all venues?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/OpenPolytechnicBITProjectGroup/Musician-Venue_App/pull/26#issuecomment-318995084>,
or mute the thread

**Comment by ttocskcaj** [comment]
I just played around with it. having `MATCH (v:Venue)-[:LIKES_GENRES]->(g:Genre) WHERE (g.name) = "" RETURN v AS venue` return nothing (sadly) But this works (just removing the WHERE part):
`MATCH (v:Venue)-[:LIKES_GENRES]->(g:Genre) RETURN v AS venue`

Edit:
Yea, but the point I'm trying to make is that searching for venues without narrowing them by genres WOULD be a valid search. Someone may want to just list all venues in Auckland, regardless of what genre the venues prefer.
The important part is trying to figure out how to change the query depending on whether or not the do want to narrow the results by genre.

**Comment by lxdnz254** [comment]
Because neo4j will iterate over relationships in our calls ie LIKES_GENRES
compute time will be quick

**Comment by ttocskcaj** [comment]
Something like this could work, perhaps?
query = "MATCH (v:Venue)-[:LIKES_GENRES]->(g:Genre)"
if(genre parameter exists){

    query += " WHERE (g.name) = {genre}"

}

if (city parameter exists){

    query += " WHERE (v.city) = {city}"

}

query += " RETURN v AS venue"

**Comment by lxdnz254** [comment]
That's sort of the road I'm heading down. I'll have to research a bit how
to do it inside the query, but for now I'll work on the conditionals on the
outside.

On 31/07/2017 8:33 PM, "Jack Scott" <notifications@github.com> wrote:

Something like this could work, perhaps?
query = "MATCH (v:Venue)-[:LIKES_GENRES]->(g:Genre)"
if(genre parameter exists){ query += " WHERE (g.name) = {genre}" } if (city
parameter exists){ query += " WHERE (v.city) = {city}" } query += " RETURN
v AS venue"

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/OpenPolytechnicBITProjectGroup/Musician-Venue_App/pull/26#issuecomment-319003286>,
or mute the thread

**Comment by lxdnz254** [comment]
pushed another commit, I think were heading down the right track now. Client $http call is to `/api/search?{type='venue | artist | fan'&genre={genre}` etc.
`route.server` sends `api/search` to `searchController.js` (serverside) 
At this stage, this is where the query filters are sorted out before choosing its search query from `db_api.js` (there is only one possibility at the moment!)

**Comment by ttocskcaj** [comment]
Yea sounds good 😄 

At this stage, "type" is probably compulsory, but we could later change that to optional as well.
E.G. to search for all artists and venues in Auckland at the same time

**Comment by lxdnz254** [comment]
Yeah, if the conditional "type" did not exist it could default to that as a
search.
Limiting user options on the client side will make our api tasks easier
also, and creates better user experience... lead them along the path.

On 31/07/2017 10:18 PM, "Jack Scott" <notifications@github.com> wrote:

Yea sounds good 😄

At this stage, "type" is probably compulsory, but we could later change
that to optional as well.
E.G. to search for all artists and venues in Auckland at the same time

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/OpenPolytechnicBITProjectGroup/Musician-Venue_App/pull/26#issuecomment-319027728>,
or mute the thread

**Comment by ttocskcaj** [comment]
I just modified a couple of minor things. Everything looks good to me.

**Comment by lxdnz254** [comment]
I've added the test for the 501 status. Everything else looks good.
Happy to merge to dev?

**Comment by lxdnz254** [comment]
Wait.... submission link is broken?

**Comment by lxdnz254** [comment]
Solved the missing $rootScope. Now everything is OK
